### ISSUE TITLE
API to send logs to core database (edited title)

### ISSUE THREAD

**Issue by GlenDC** [ulixee/hero#157]
As far as I understand the Secret-Agent code, and I mind you that I'm still in the process of fully grasping the entire codebase,
logs from all log levels end up at:

https://github.com/ulixee/secret-agent/blob/db3f3535636707ed2fd9c7f5dcc29f7d4bd07cc7/commons/Logger.ts#L47-L103

The problem with this is that from a light client there is no way for me to really capture those logs and send them back for my own purposes. Let me show you, if you will, how I currently make use of Secret-Agent.

I create my core as follows:

```typescript
import Core from '@secret-agent/core';

await Core.start({
        coreServerPort: port,
});
```

For each use of the core I create a new agent as follows:

```typescript
this.agent = new Agent({
        name: 'brorun-playground',
        userAgent,
        showReplay: false,
        connectionToCore: new RemoteConnectionToCore({
                host: coreAddress,
        }),
});
```

And then I make use of the agent directly (e.g. to go to a page, wait for stuff, interact with the dom) by using that `this.agent`.

At the end I also close the agent as follows:

```typescript
await this.agent.close();
```

Now the problem is that the Core, which I imagine is started as a child process (haven't checked to be sure) is doing this logging but as we can see from the Logging code it logs directly to the console. The brute-force way would be to capture the logs, but this would not restrict itself to logs just for that one client but capture it all. I could than filter on sessionID but that seems dirty?

What would you recommend on getting these logs per agent? I don't think there is a way currently to do so, but would you be open to think about how we could handle that? Would it be something you would be willing to support (with the default still being that we just log it all directly to the console. E.g. a first initial kick-off for the proposal could be:

- we attach the logger to an active session;
- by default the logger's internal stream is streaming it to `console.log` (stdout?!)
- however we could allow this to be buffered instead, and provide a command to the agent to get the `logs`, in a similar fashion that the agent can get the entire page's content.

---

Related to this question I also wonder what one can really do with `await agent.sessionId`. It seems to me that he code to actually get the replay via the official code is not really exposed (e.g. `getSession` etc). Is one expected to work directly with the undocumented SQL files? Not that it would help me for these purposes as it doesn't seems that the logs are stored in there?

**Comment by GlenDC** [comment]
This approach I am still taking more or less ,where I read my logs via SQLite and than send them. Would be nice to somehow get this integrated somehow. Same for the cleaning process. I feel that could be 1 epic. I mean I get the use case where you do want to keep it, but might be cool to offer the featue to have it.

I think we can just implement this as a separate ulixee server sub module or w/e which allows you to manage the SQL Session Data.
- wipe a session
- retrieve bits of the session, ... 
- etc

For the retrieving bits we can for now keep it as simple as only being able to retrieve the logs. This way you do not need to expose the entire data structure of the entire DB, yet.

For the wiping we can already allow to wipe an entire session, as it doesn't require you to expose anything, it's essentail a single blackbox function.

So in short, I think we can provide submodule somewhere for this which provides a class that you can use to interact with the DB. 

How does that sound @blakebyrnes ?

**Comment by blakebyrnes** [comment]
I think we're talking about two pretty different things here. I was imagining an ability to simply "add" to the logs database. We do have an API layer in Core already that exposes a good bit of this kind of information. Maybe exposing logs in there is the right approach for that piece? It's not a documented feature at the moment ... so we'd need to document it if we decide this is going to become first class. I'm on the fence about making that a "stable" api that we stick to though. It continues to move around a good bit..

Your second part about managing sessions and removing the databases is a very valid item we get a lot of requests for. I think a pattern for this would be very nice... however, I'd have to give it a good bit of thought. I think it might be simply subscribing to a 'session-complete' event in Core that exposes a db path, or something to that effect. Then the process can choose whether they want to zip it, send it off, or just wipe it. I think this should probably go in a different ticket (or the "log to session database" api can be a different ticket)

**Comment by blakebyrnes** [comment]
See #166 for a feature supporting shutdown
